File: Impostor/CollageViewController.m
//
//  CollageViewController.m
//  Impostor
//
//  Created by William Entriken on 2/19/14.
//  Copyright (c) 2014 William Entriken. All rights reserved.
//

#import "CollageViewController.h"
#import "ImpostorGameModel.h"
#import "GAIDictionaryBuilder.h"
#import <AVFoundation/AVAudioPlayer.h>
#import <AudioToolbox/AudioToolbox.h>
#import "GameConfigurationViewController.h"
#import "CachedPersistentJPEGImageStore.h"

@interface CollageViewController () <UICollectionViewDataSource>
@property (nonatomic) ImpostorGameModel *game;
@property (nonatomic, strong) AVAudioPlayer *audioPlayer;
@end

@implementation CollageViewController

- (ImpostorGameModel *)game
{
    if (!_game)
        _game = [ImpostorGameModel game];
    return _game;
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    self.screenName = @"CollageViewController";
    
    NSURL *url = [[NSBundle mainBundle] URLForResource:@"camera" withExtension:@"mp3"];
    self.audioPlayer = [[AVAudioPlayer alloc] initWithContentsOfURL:url error:nil];
    [self performSelector:@selector(flashScreen) withObject:nil afterDelay:1.0];
    [self performSelector:@selector(showShareBox) withObject:nil afterDelay:1.8];
    
    GameConfigurationViewController *root = (GameConfigurationViewController *)self.navigationController.viewControllers.firstObject;
    [root fadeOutMusic];
}

- (void)flashScreen
{
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.audioPlayer play];
        UIView *flashView = [[UIView alloc] initWithFrame:self.view.frame];
        [flashView setBackgroundColor:[UIColor whiteColor]];
        [[[self view] window] addSubview:flashView];
        
        [UIView animateWithDuration:0.7f
                         animations:^{
                             [flashView setAlpha:0.f];
                         }
                         completion:^(BOOL finished){
                             [flashView removeFromSuperview];
                         }
         ];
    });
}

- (void)willRotateToInterfaceOrientation:(UIInterfaceOrientation)toInterfaceOrientation duration:(NSTimeInterval)duration
{
    [super willRotateToInterfaceOrientation:toInterfaceOrientation duration:duration];
    [self.view setNeedsLayout];
}

#pragma mark - UICollectionViewDataSource

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return self.game.numberOfPlayers;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    UICollectionViewCell *cell = [self.playerPhotoCollectionView dequeueReusableCellWithReuseIdentifier:@"playerCell" forIndexPath:indexPath];
    UIImage *photo = [UIImage imageNamed:@"defaultHeadshot.png"];
    UIImageView *imageView = (UIImageView *)[cell viewWithTag:1];
    imageView.image = photo;
    
    NSString *imageName = [NSString stringWithFormat:@"%ld", (long)indexPath.row];
    photo = [[CachedPersistentJPEGImageStore sharedStore] imageWithName:imageName];
    if (photo)
        imageView.image = photo;

#define ARC4RANDOM_MAX      0x100000000
    double val = ((double)arc4random() / ARC4RANDOM_MAX);
    double rotate = M_PI*30/180*(val-0.5);

    imageView.transform = CGAffineTransformMakeRotation(rotate);
    CAKeyframeAnimation *bounceAnimation = [CAKeyframeAnimation animationWithKeyPath:@"transform.rotation.z"];
    bounceAnimation.values = [NSArray arrayWithObjects:@(0), @(rotate*0.5), @(-rotate*2), @(rotate), nil];
    bounceAnimation.duration = ((double)arc4random() / ARC4RANDOM_MAX);
    bounceAnimation.removedOnCompletion = YES;
    bounceAnimation.repeatCount = 0;
    bounceAnimation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    [imageView.layer addAnimation:bounceAnimation forKey:@"spin"];
    
    [self makeItBounce:imageView];
    return cell;
}

- (void)viewDidLayoutSubviews
{
    NSInteger cellMaxSide = 0, rows=9999, cols=9999;
    do {
        // if your cellWidth != cellHeight, switch here to set the larger then smaller of the two
        cols = floorf((self.playerPhotoCollectionView.frame.size.width+10)/((cellMaxSide+1)+10));
        rows = floorf((self.playerPhotoCollectionView.frame.size.height+10)/((cellMaxSide+1)+10));
    } while ((rows * cols >= self.game.numberOfPlayers) && ++cellMaxSide);
    CGFloat dim = MIN(cellMaxSide, cellMaxSide);
    
    [(UICollectionViewFlowLayout *)self.playerPhotoCollectionView.collectionViewLayout setItemSize:CGSizeMake(dim,dim)];
//    [self.playerPhotoCollectionView reloadData];
}

- (void)makeItBounce:(UIView *)view
{
    CAKeyframeAnimation *bounceAnimation = [CAKeyframeAnimation animationWithKeyPath:@"transform.scale"];
    bounceAnimation.values = [NSArray arrayWithObjects:[NSNumber numberWithFloat:1], [NSNumber numberWithFloat:1.2], nil];
    bounceAnimation.duration = 0.15;
    bounceAnimation.removedOnCompletion = NO;
    bounceAnimation.repeatCount = 2;
    bounceAnimation.autoreverses = YES;
    bounceAnimation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    [view.layer addAnimation:bounceAnimation forKey:@"bounce"];
}

- (void)showShareBox
{
    // Create the item to share (in this example, a url)
    NSString *appName = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleDisplayName"];
    NSURL *url = [NSURL URLWithString:@"https://itunes.apple.com/us/app/whos-the-impostor/id784258202"];
    NSString *title = [NSString stringWithFormat:NSLocalizedString(@"I am playing the party game %@", @"Text for sharing on social network"), appName];
    UIImage* image = nil;
    UIGraphicsBeginImageContext(self.view.frame.size);
    [self.view.layer renderInContext: UIGraphicsGetCurrentContext()];
    image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    NSArray *itemsToShare = @[image, title, url];
    UIActivityViewController *activityVC = [[UIActivityViewController alloc] initWithActivityItems:itemsToShare applicationActivities:nil];
    activityVC.excludedActivityTypes = @[UIActivityTypePrint, UIActivityTypeAssignToContact];
    
    [activityVC setCompletionHandler:^(NSString *activityType, BOOL completed) {
        id<GAITracker> tracker = [[GAI sharedInstance] defaultTracker];
        [tracker send:[[GAIDictionaryBuilder createEventWithCategory:@"Gameplay"
                                                              action:@"Gameover"
                                                               label:@"SharedCollage"
                                                               value:@(completed)] build]];
        [self.navigationController popToRootViewControllerAnimated:YES];
    }];
    [self presentViewController:activityVC animated:YES completion:nil];
}

@end

File: Impostor/CollageViewController.swift
//
//  CollageViewController.swift
//  Impostor
//
//  Created by Full Decent on 6/20/16.
//  Copyright © 2016 William Entriken. All rights reserved.
//

import UIKit
import AVFoundation
import AudioToolbox

class CollageViewController: UIViewController {
    fileprivate var game = ImpostorGameModel.game
    fileprivate var audioPlayer: AVAudioPlayer = {
        let url = Bundle.main.url(forResource: "camera", withExtension: "mp3")!
        return try! AVAudioPlayer(contentsOf: url)
    }()
    
    @IBOutlet weak var playerPhotoCollectionView: UICollectionView!

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
    }
    
    @objc func flashScreen() {
        DispatchQueue.main.async(execute: {
            self.audioPlayer.play()
            let flashView = UIView(frame: self.view.frame)
            flashView.backgroundColor = UIColor.white
            self.view!.window!.addSubview(flashView)
            UIView.animate(withDuration: 0.7, animations: {
                flashView.alpha = 0.0
                }, completion: {
                    _ in
                    flashView.removeFromSuperview()
            })
        })
    }
    
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        view!.setNeedsLayout()
    }
    
    @IBAction func wantToShare() {
        self.perform(#selector(CollageViewController.flashScreen), with: nil, afterDelay: 0.1)
        self.perform(#selector(CollageViewController.showShareBox), with: nil, afterDelay: 0.4)
    }
    
    @objc func showShareBox() {
        // Create the item to share (in this example, a url)
        // http://stackoverflow.com/a/25837053/300224
        let mainBundle = Bundle.main
        let displayName = mainBundle.object(forInfoDictionaryKey: "CFBundleDisplayName") as? String
        let name = mainBundle.object(forInfoDictionaryKey: kCFBundleNameKey as String) as? String
        let appName = displayName ?? name ?? "Unknown"
        let url = URL(string: "https://itunes.apple.com/us/app/whos-the-impostor/id784258202")!
        let title = String(format: NSLocalizedString("I am playing the party game %@", comment:"Number of players in the game"), appName)
        UIGraphicsBeginImageContext(self.view.frame.size)
        for subView in self.view.subviews {
            if subView is UIButton {
                subView.isHidden = true
            }
        }
        view.layer.render(in: UIGraphicsGetCurrentContext()!)
        for subView in self.view.subviews {
            if subView is UIButton {
                subView.isHidden = false
            }
        }
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        let itemsToShare: [AnyObject] = [image!, title as AnyObject, url as AnyObject]
        let activityVC: UIActivityViewController = UIActivityViewController(activityItems: itemsToShare, applicationActivities: nil)
        activityVC.excludedActivityTypes = [UIActivity.ActivityType.print, UIActivity.ActivityType.assignToContact]
        activityVC.completionWithItemsHandler = {
            activityType, completed, returnedItems, activityError in
        }
        self.present(activityVC, animated: true, completion: nil)
    }

    @IBAction func returnToMainScreen() {
        self.navigationController!.popToRootViewController(animated: true)
    }
}

extension CollageViewController: UICollectionViewDataSource {
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.game.numberOfPlayers
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = playerPhotoCollectionView.dequeueReusableCell(withReuseIdentifier: "playerCell", for: indexPath)
        let imageView = cell.viewWithTag(1) as! UIImageView
        let imageName = "\(Int((indexPath as NSIndexPath).row))"
        if let photo = CachedPersistentJPEGImageStore.sharedStore.imageWithName(imageName) {
            imageView.image = photo
        } else {
            imageView.image = UIImage(named: "defaultHeadshot.png")!
        }
        
        let degrees = Double.random(in: 15...15)
        let radians = CGFloat(degrees * Double.pi / 180)
        imageView.transform = CGAffineTransform(rotationAngle: radians)
        let bounceAnimation: CAKeyframeAnimation = CAKeyframeAnimation(keyPath: "transform.rotation.z")
        bounceAnimation.values = [0, radians * 0.5, -radians * 2, radians]
        bounceAnimation.duration = Double.random(in: 0...1)
        bounceAnimation.isRemovedOnCompletion = true
        bounceAnimation.repeatCount = 0
        bounceAnimation.timingFunction = CAMediaTimingFunction(name: CAMediaTimingFunctionName.easeOut)
        imageView.layer.add(bounceAnimation, forKey: "spin")
        self.makeItBounce(imageView)
        
        return cell
    }
    
    override func viewDidLayoutSubviews() {
        var highestWorking = 0
        var lowestNotWorking = 9999
        while lowestNotWorking > highestWorking + 1 {
            let currentTest = (highestWorking + lowestNotWorking) / 2
            let cols = Int(playerPhotoCollectionView.frame.size.width + 10) / (currentTest + 10)
            let rows = Int(playerPhotoCollectionView.frame.size.height + 10) / (currentTest + 10)
            if cols * rows >= game.numberOfPlayers {
                highestWorking = currentTest
            } else {
                lowestNotWorking = currentTest
            }
        }
        let size = CGSize(width: CGFloat(highestWorking), height: CGFloat(highestWorking))
        (self.playerPhotoCollectionView.collectionViewLayout as! UICollectionViewFlowLayout).itemSize = size
        super.viewDidLayoutSubviews()
    }
    
    func makeItBounce(_ view: UIView) {
        let bounceAnimation: CAKeyframeAnimation = CAKeyframeAnimation(keyPath: "transform.scale")
        bounceAnimation.values = [1.0, 1.2]
        bounceAnimation.duration = 0.15
        bounceAnimation.isRemovedOnCompletion = false
        bounceAnimation.repeatCount = 2
        bounceAnimation.autoreverses = true
        bounceAnimation.timingFunction = CAMediaTimingFunction(name: CAMediaTimingFunctionName.easeOut)
        view.layer.add(bounceAnimation, forKey: "bounce")
    }
}
